contribute/design/1.x/replica_scaleup/dataplane/20190910-replica-scaleup.md

```diff
@@ -0,0 +1,278 @@
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Current Implementation -- Config](#current-implementation----config)
+      * [Current Implementation](#current-implementation)
+      * [Shortcomings with Current Implementation](#shortcomings-with-current-implementation)
+    * [Proposed Implementation](#proposed-implementation)
+      * [Adding new replica](#adding-new-replica-workflow)
+      * [Replacing non-existing replica](#replacing-non-existing-replica-workflow)
+    * [High Level Design](#high-level-design)
+      * [Components Interaction](#components-interaction)
+      * [Sample YAMLs](#sampleyamls)
+      * [Notes](#notes)
+    * [Low Level Design](#low-level-design)
+      * [Replica Connection](#replica-connection)
+      * [Maintaining Known Replicas List](#maintaining-known-replicas-list)
+      * [Notes](#notes)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+* [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This proposal includes design for cStor data plane to allow adding new replicas
+to it so that replication factor of the volume will be increased.
+
+## Motivation
+
+There are cases arising where OpenEBS user lost multiple copies of data, and,
+working with available single copy.
+This proposal is to enable him to add more replicas to it again.
+
+### Goals
+
+- Increase ReplicationFactor of volume by adding replicas
+- Identify replicas which are allowed to connect to target
+
+### Non-Goals
+
+- Scaling down replicas
+- Add replicas in GitOps model
+- Workflow to replace non-existing replica
+- operator that detects the need to increase ReplicationFactor
+- Allow adding replicas to replace particular replica (volume distribution case)
+
+## Proposal
+
+### User Stories
+
+#### Add new replica
+As an OpenEBS user, I should be able to add replicas to my volume.
+
+#### Replace non-existing replica
+As an OpenEBS user, I should be able to replace non-existing replica with new
+replica.
+
+### Implementation Details/Notes/Constraints
+
+Currently, cstor-istgt reads replication related configuration, i.e.,
+ReplicationFactor (RF) and ConsistencyFactor (CF), from istgt.conf file.
+cstor-volume-mgmt is the container that fills the RF and CF details into
+istgt.conf file by reading from CStorVolume CR.
+
+Another property that cstor-istgt uses is 'quorum'. This property is set at
+replica, and, replica sends this to istgt during handshake phase.
+This property can take values of 'on' and 'off'.
+'on' means that data written to this replica is available for reads.
+'off' means that data written to this replica is lost.
+cstor-pool-mgmt is the container that sets quorum property at replica.
+It sets property as 'on' if replica is created first time related to that CVR.
+Otherwise, it sets to 'off'.
```

Comment:
Needs some more clarification. 
- First time creation, q=on
- If it is recreated due to underlying pool is recreated, q=off. 

Is there a third case?  What is the state when replica/pool is restarted and is not in sync with data. Will it come up q=off or q=on?

How does the cstor-istgt use this value during the handshake phase?

Reply:
There is no third case for cstor-pool-mgmt @kmova. Only 2 cases - fresh creation of replica for that CVR CR or not.
For the case you mentioned, cstor-pool-mgmt will not come into picture. As pool gets imported, the 'quorum' value which was already set on the replica will be sent to target in handshake process.
At cstor-istgt, if quorum is off for a replica, that replica won't participate in IO related consistency factor checks. It just there to rebuild missing data while its taking ongoing IOs.
If quorum is on, cstor-istgt adds it to the replicas which decides the fate of IOs based on their success/failure on these quorum replicas.
I will update above info in OEP.